sklearn/datasets/openml.py

```diff
@@ -0,0 +1,306 @@
+import json
+import sys
+import os
+from os.path import join, exists
+from warnings import warn
+
+try:
+    # Python 2
+    from urllib2 import urlopen
+except ImportError:
+    # Python 3+
+    from urllib.request import urlopen
+
+
+import numpy as np
+
+from .base import get_data_home
+from ..externals._liacarff.arff import loads
+from ..externals.joblib import Memory
+from ..externals.six.moves.urllib.error import HTTPError
+from ..utils import Bunch
+
+_SEARCH_NAME = "https://openml.org/api/v1/json/data/list/data_name/{}/limit/1"
+_DATA_INFO = "https://openml.org/api/v1/json/data/{}"
+_DATA_FEATURES = "https://openml.org/api/v1/json/data/features/{}"
+
+
+def _openml_fileid_url(file_id):
+    return "https://www.openml.org/data/v1/download/{}/".format(file_id)
+
+
+def _convert_arff_data(arff_data):
+    """
+    converts the arff object into the appropriate matrix type (now: np.ndarray,
+    later also: scipy.sparse.csr_matrix) based on the 'data part' (i.e., in the
+    liac-arff dict, the object from the 'data' key)
+
+    Parameters
+    ----------
+    arff_data : list or dict
+        as obtained from liac-arff object
+
+    returns : np.ndarray (or later also: scipy.sparse.csr_matrix)
+    """
+    if isinstance(arff_data, list):
+        X = np.array(arff_data, dtype=object)
```

Comment:
If we care about performance, I think we should not yet convert it to an object array here, but only do that once we know the dtype of the final array (eg if only numeric, we can directly cast the full array to float). 
But not sure if this is important, downloading the data will typically take a lot more time than this conversion.

Reply:
Agreed with your point, but please keep in mind the following: We can obtain data in both sparse and dense format (currently all is dense in liac-arff). With this fn we convert it into an array (dense or later also sparse) and we can follow the array api, rather than having to work with two different api's. IMO this makes the code easier to read and maintain. If you disagree we can of course refactor the code.